### PR TITLE
Refactor validator windowing

### PR DESCRIPTION
**Pull request by pgarg66**
- a unit test for windowing functions
- issue #857

### CODE REVIEW

File: src/streamer.rs [hunk 1]
```diff
@@ -442,6 +442,39 @@ fn process_blob(
     }
 }
 
+fn validate_blob_against_window(
+    debug_id: u64,
+    pix: u64,
+    consumed: u64,
+    received: &mut u64,
+) -> bool {
```
Comment (garious) [inline src/streamer.rs]: Return `Option<u64>`?

Comment (pgarg66) [reply src/streamer.rs]: with u64 carrying received?

Comment (pgarg66) [reply src/streamer.rs]: @garious , wondering how Option helps here. Is this to differentiate between different error conditions? Or something more?

Comment (garious) [reply src/streamer.rs]: Yes, with u64 carrying received. It's more idiomatic Rust to avoid `mut`. You'll see those tests get shorter.

**Comment by pgarg66** [comment]
@garious , please see the latest patch. It simplified the tests, as you had mentioned.

File: src/streamer.rs [hunk 2]
```diff
@@ -442,6 +442,39 @@ fn process_blob(
...
+    // probably from a repair window request
+    if pix < consumed {
+        debug!(
+            "{:x}: received: {} but older than consumed: {} skipping..",
+            debug_id, pix, consumed
+        );
+        return None;
+    }
+
+    if pix > received {
```
Comment (garious) [inline src/streamer.rs]: `Some(cmp::max(pix, received))`

File: src/streamer.rs [hunk 3]
```diff
@@ -489,25 +522,15 @@ fn recv_window(
...
-        // Got a blob which has already been consumed, skip it
-        // probably from a repair window request
-        if pix < *consumed {
-            debug!(
-                "{:x}: received: {} but older than consumed: {} skipping..",
-                debug_id, pix, *consumed
-            );
+
+        let result = validate_blob_against_window(debug_id, pix, *consumed, *received);
+        if result.is_none() {
```
Comment (garious) [inline src/streamer.rs]: Can you use a `match` here? Under the hood, both `is_none()` and `result.unwrap()` below would use a match. I'm guessing clippy is about to suggest this as well. Let's see!

File: src/streamer.rs [hunk 4]
```diff
@@ -1107,4 +1131,17 @@ mod test {
             WINDOW_SIZE + 9
         );
     }
+
+    #[test]
+    pub fn validate_blob_against_window_test() {
+        let recvd: u64 = 100;
+        assert!(validate_blob_against_window(0, 90 + WINDOW_SIZE, 90, recvd).is_none());
```
Comment (garious) [inline src/streamer.rs]: Can you use `assert_eq!` in all of these? It'll give you better error messages.

File: src/streamer.rs [hunk 5]
```diff
@@ -442,6 +442,39 @@ fn process_blob(
...
 
+fn validate_blob_against_window(
+    debug_id: u64,
+    pix: u64,
+    consumed: u64,
+    received: u64,
+) -> Option<u64> {
+    // Prevent receive window from running over
+    if pix >= consumed + WINDOW_SIZE {
+        debug!(
```
Comment (garious) [inline src/streamer.rs]: Does it bother anyone that this information is lost if not logging in debug mode? I see the original code did this, so I'm not going to ask you to clean that up here, but could you add a ticket to add an error type and change this function from `Option<u64>` to `Result<u64>`?

File: src/streamer.rs (hunk 3 above)
Comment (pgarg66) [reply src/streamer.rs]: Funny, as I used match initially, and then changed it to this. I'll change it to match.

File: src/streamer.rs (hunk 4 above)
Comment (pgarg66) [reply src/streamer.rs]: Sure. I'll change it.

File: src/streamer.rs (hunk 2 above)
Comment (pgarg66) [reply src/streamer.rs]: Cool.

File: src/streamer.rs (hunk 5 above)
Comment (pgarg66) [reply src/streamer.rs]: Sure, will do.

**Comment by pgarg66** [comment]
@garious , this should address most of the comments.

**Pull request merged**